File: CodingInterviewsPythonSolutions/205. Isomorphic Strings.py
class Solution:
    def canonical(self, s: str) -> List[int]:
        myMap = {}
        canonicalIndex = 0
        result = []
        
        for ch in s:
            if ch not in myMap:
                myMap[ch] = canonicalIndex
                canonicalIndex += 1
            result.append(myMap[ch])
        
        return result
    
    def isIsomorphic(self, s: str, t: str) -> bool:
        return self.canonical(s) == self.canonical(t)

File: CodingInterviewsJavaSolutions/205. Isomorphic Strings.java
class Solution {
    public int[] canonical(String s) {
        HashMap<Character, Integer> myMap = new HashMap<>();
        int canonicalIndex = 0;
        int[] result = new int[s.length()];
        
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (!myMap.containsKey(ch)) {
                myMap.put(ch, canonicalIndex);
                canonicalIndex++;
            }
            result[i] = myMap.get(ch);
        }
        
        return result;
    }
    
    public boolean isIsomorphic(String s, String t) {
        int[] sCanonical = canonical(s);
        int[] tCanonical = canonical(t);
        
        if (sCanonical.length != tCanonical.length) {
            return false;
        }
        
        for (int i = 0; i < sCanonical.length; i++) {
            if (sCanonical[i] != tCanonical[i]) {
                return false;
            }
        }
        
        return true;
    }
}
